### PR TITLE
fix: fixed data race in introduce protocol

### PR DESCRIPTION
**Pull request by soluchok**
* data race
* goroutines leak (in tests)

Part of https://github.com/hyperledger/aries-framework-go/issues/269

Signed-off-by: Andrii Soluk <isoluchok@gmail.com>

### CODE REVIEW

**Comment by rolsonquadras** [comment]
just a suggestion - better to have a specific commit title for rather than generic one (feat: Introduce protocol server implementation). While going through commit history, the title doesn't give specific info and looks [same](https://github.com/hyperledger/aries-framework-go/commits?author=soluchok).

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -59,6 +59,11 @@ jobs:
     timeout-minutes: 10
     runs-on: ubuntu-18.04
     steps:
+      - name: Setup Go 1.13
```
Comment (rolsonquadras) [inline .github/workflows/build.yml]: Looks like this commit contains same code as in PR - https://github.com/hyperledger/aries-framework-go/pull/658/files.

File: pkg/didcomm/protocol/introduce/persistence.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+		return nil, err
+	}
+
+	return rec, nil
+}
+
+// SaveRecord saves data
+// this function usually is executed by the introduce client
+// NOTE: if the record already exists in the store it will rewrite it
+func (p *Persistence) SaveRecord(thID string, rec *Record) error {
```
Comment (rolsonquadras) [inline pkg/didcomm/protocol/introduce/persistence.go]: Ideally client shouldn't save protocol related records and it should be done through service pkg (knowing of lower layer primitives like threadID). What's the use case here ?

File: pkg/didcomm/protocol/introduce/persistence.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	return &Persistence{store: store}
+}
+
+// GetRecord returns data by given threadID
+// this function should be used in introduce service
+// state machine behavior depends on this record's data
+func (p *Persistence) GetRecord(thID string) (*Record, error) {
+	var rec *Record
+	if err := p.getData(recordKey(thID), &rec); err != nil {
+		return nil, err
```
Comment (rolsonquadras) [inline pkg/didcomm/protocol/introduce/persistence.go]: May be wrapping error message here would help with error message stack.

File: .github/workflows/build.yml (hunk 1 above)
Comment (soluchok) [reply .github/workflows/build.yml]: I mentioned in the description `Depends on PR #658`

File: pkg/didcomm/protocol/introduce/persistence.go (hunk 2 above)
Comment (soluchok) [reply pkg/didcomm/protocol/introduce/persistence.go]: the client has a bunch of data needed for service, we have similar logic in `didexchange` client e.g `connectionStore: didexchange.NewConnectionRecorder(transientStore, store),`

Comment (rolsonquadras) [reply pkg/didcomm/protocol/introduce/persistence.go]: Yes, but not save record - https://github.com/hyperledger/aries-framework-go/blob/master/pkg/didcomm/protocol/didexchange/persistence.go#L195

Comment (soluchok) [reply pkg/didcomm/protocol/introduce/persistence.go]: `Record` it is just a name :) which keeps dependency needed for the state machine https://github.com/hyperledger/aries-framework-go/pull/711/files#diff-2be7cff09231a8b321a1f1aed45a4913R24-R28

Comment (troyronda) [reply pkg/didcomm/protocol/introduce/persistence.go]: @fqutishat @llorllale @rolsonquadras @sandrask

Comment (troyronda) [reply pkg/didcomm/protocol/introduce/persistence.go]: Is the issue that the method should be named to something more specific?
@soluchok @rolsonquadras

Comment (rolsonquadras) [reply pkg/didcomm/protocol/introduce/persistence.go]: @troyronda My comment was about the 2nd item that you pointed out. ie - service should be persisting thread related data. ie. un-export the function.

I see the db operations from different layers as follows.
1. Service layer - handle db operations for did-comm message related data. ie, based on threadID or messageID etc.
2. Client layer - handle db operation for agent/application specific data. ie, based on connectionID(which is specific to the agent)

Note - service layer can handle pt#2 as well.

**Pull request merged**